core/arns/src/main/java/software/amazon/awssdk/arns/Arn.java

```diff
@@ -0,0 +1,372 @@
+            throw new IllegalArgumentException("Malformed Arn - no service specified");
+        }
+        String service = arn.substring(partitionColonIndex + 1, serviceColonIndex);
+
+        int regionColonIndex = arn.indexOf(':', serviceColonIndex + 1);
+        if (regionColonIndex < 0) {
+            throw new IllegalArgumentException("Malformed Arn - no AWS region partition specified");
+        }
+        String region = arn.substring(serviceColonIndex + 1, regionColonIndex);
+
+        int accountColonIndex = arn.indexOf(':', regionColonIndex + 1);
+        if (accountColonIndex < 0) {
+            throw new IllegalArgumentException("Malformed Arn - no AWS account specified");
+        }
+        String accountId = arn.substring(regionColonIndex + 1, accountColonIndex);
+
+        String resource = arn.substring(accountColonIndex + 1);
+        if (resource.isEmpty()) {
+            throw new IllegalArgumentException("Malformed Arn - no resource specified");
+        }
+
+        return Arn.builder()
+                  .partition(partition)
+                  .service(service)
+                  .region(region)
+                  .accountId(accountId)
+                  .resource(resource)
+                  .build();
+    }
+
+    @Override
+    public String toString() {
+        return "arn:"
+               + this.partition
+               + ":"
+               + this.service
+               + ":"
+               + region
+               + ":"
+               + this.accountId
+               + ":"
+               + this.resource;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+
+        Arn arn = (Arn) o;
+
+        if (!Objects.equals(partition, arn.partition)) {
+            return false;
+        }
+        if (!Objects.equals(service, arn.service)) {
+            return false;
+        }
+        if (!Objects.equals(region, arn.region)) {
+            return false;
+        }
+        if (!Objects.equals(accountId, arn.accountId)) {
+            return false;
+        }
+        if (!Objects.equals(resource, arn.resource)) {
+            return false;
+        }
+        return Objects.equals(arnResource, arn.arnResource);
+    }
+
+    @Override
+    public int hashCode() {
+        int hash = partition.hashCode();
+        hash = 31 * hash + service.hashCode();
+        hash = 31 * hash + region.hashCode();
+        hash = 31 * hash + accountId.hashCode();
+        hash = 31 * hash + resource.hashCode();
+        return hash;
+    }
+
+    @Override
+    public Builder toBuilder() {
+        return builder().accountId(accountId)
+                        .partition(partition)
+                        .region(region)
+                        .resource(resource)
+                        .service(service)
+            ;
+    }
```

Comment:
There is no test coverage for this code.

Reply:
added tests